usl_models/usl_models/flood_ml/model.py

```diff
@@ -48,72 +50,102 @@ def _build_model(self) -> tf.keras.Model:
+        if tf.rank(data.temporal) == 3:  # windowed: (B, T_max, m)
+            assert data.temporal.shape[-1] == self._model_params.m_rainfall, (
+                "Mismatch between the temporal data window size "
+                f"({data.temporal.shape[-1]}) and the expected window size "
+                f"(m = {self._model_params.m_rainfall})."
+            )
+        else:
+            full_temp_input = data_utils.temporal_window_view(
+                data.temporal, self._model_params.m_rainfall
+            )
+            data = replace(data, temporal=full_temp_input)
 
         # We assume that, if provided, this input is a *single* flood map.
-        st_input = inputs.get("spatiotemporal")
+        st_input = data.spatiotemporal
         if st_input is None:
-            st_shape = geo_input.shape[:3] + [1]
+            st_shape = data.geospatial.shape[:3] + [1]
             st_input = tf.zeros(st_shape)
 
-        return {
-            "geospatial": geo_input,
-            "temporal": full_temp_input,
-            "spatiotemporal": st_input,
-        }
+        data = replace(data, spatiotemporal=st_input)
+        return data
 
-    def train(
-        self,
-        train_data: dict[str, tf.Tensor],
-        train_labels: tf.Tensor,
-        n_predictions=1,
-    ) -> tf.keras.callbacks.History:
-        """Trains the model.
+    def _model_fit(self, data: FloodModelData) -> tf.keras.callbacks.History:
+        """Fits the model on a single batch of FloodModelData.
 
         Args:
-            train_data: A dictionary of training inputs.
-            train_labels: A tensor of labels on which to fit the model.
-            n_predictions: The number of predictions to make, synonymous with
-                the storm duration.
+            data: A FloodModelData object.
 
-        Returns:
-            A History object containing the record of training and, if applicable,
-            validation loss and metrics.
+        Returns: A History object containing the training and validation loss
+            and metrics.
         """
-        train_inputs = self._validate_and_preprocess_inputs(train_data)
-        self._model.set_n_predictions(n_predictions)
+        inputs = {
+            "spatiotemporal": data.spatiotemporal,
+            "geospatial": data.geospatial,
+            "temporal": data.temporal,
+        }
+        self._model.set_n_predictions(data.storm_duration)
         history = self._model.fit(
-            train_inputs,
-            train_labels,
+            inputs,
+            data.labels,
             batch_size=self._model_params.batch_size,
             epochs=self._model_params.epochs,
             validation_split=0.2,
         )
         return history
 
+    def train(
+        self,
+        data: list[FloodModelData],
+    ) -> list[tf.keras.callbacks.History]:
+        """Trains the model.
+
+        Args:
+            data: A list of FloodModelData objects. Objects should be grouped by storm
```

Comment:
'grouped by storm duration' in what sense? Partly a data structure questions -- should it be
```python
    storm_duration: int
    geospatial: Iterable[tf.Tensor]
```
to group all tensors with the same storm duration?

Reply:
This is a poor description on my part. Each `FloodModelData` instance is associated with a single storm duration. The list is to allow passing in data for multiple storm durations. In other words, data for different storm durations must be passed in as different `FloodModelData` objects.

Reply:
Okay, cool, what you describe here in the comment box sounds good -- might make the language in the docstring as well.